[clang] Introduce diagnostics suppression mappings

kadircet opened the pull request:
This implements
https://discourse.llvm.org/t/rfc-add-support-for-controlling-diagnostics-severities-at-file-level-granularity-through-command-line/81292.

Users now can suppress warnings for certain headers by providing a
mapping with globs, a sample file looks like:
```
[unused]
src:*
src:*clang/*=emit
```

This will suppress warnings from `-Wunused` group in all files that
aren't under `clang/` directory. This mapping file can be passed to
clang via `--warning-suppression-mappings=foo.txt`.

At a high level, mapping file is stored in DiagnosticOptions and then
processed with rest of the warning flags when creating a
DiagnosticsEngine. This is a functor that uses SpecialCaseLists
underneath to match against globs coming from the mappings file.

This implies processing warning options now performs IO, relevant
interfaces are updated to take in a VFS, falling back to RealFileSystem
when one is not available.

kadircet committed:
[clang] Introduce diagnostics suppression mappings

This implements
https://discourse.llvm.org/t/rfc-add-support-for-controlling-diagnostics-severities-at-file-level-granularity-through-command-line/81292.

Users now can suppress warnings for certain headers by providing a
mapping with globs, a sample file looks like:
```
[unused]
src:*
src:*clang/*=emit
```

This will suppress warnings from `-Wunused` group in all files that
aren't under `clang/` directory. This mapping file can be passed to
clang via `--warning-suppression-mappings=foo.txt`.

At a high level, mapping file is stored in DiagnosticOptions and then
processed with rest of the warning flags when creating a
DiagnosticsEngine. This is a functor that uses SpecialCaseLists
underneath to match against globs coming from the mappings file.

This implies processing warning options now performs IO, relevant
interfaces are updated to take in a VFS, falling back to RealFileSystem
when one is not available.

kadircet committed:
clang-format

ilya-biryukov commented on clang/include/clang/Basic/DiagnosticOptions.h:
```diff
@@ -108,6 +108,9 @@ class DiagnosticOptions : public RefCountedBase<DiagnosticOptions>{
   /// The file to serialize diagnostics to (non-appending).
   std::string DiagnosticSerializationFile;
 
+  /// File that defines suppression mappings.
```
NIT: diagnostic suppression mappings?

I can imagine being confused had I not had this context ("what are supressions and why we need to map them?")

kadircet replied:
I was dropping that mostly because this is in `DiagnosticOptions`, but adding those to names to be consistent with others here.

ilya-biryukov commented on clang/lib/Basic/DiagnosticIDs.cpp:
```diff
@@ -575,25 +575,42 @@ DiagnosticIDs::getDiagnosticSeverity(unsigned DiagID, SourceLocation Loc,
 
+  // Rest of the mappings are only applicable for diagnostics associated with a
+  // SourceLocation, bail out early for others.
+  if (!Diag.hasSourceManager())
```
NIT: maybe split into a separate review and land right away?

This looks pretty orthogonal to the rest of the patch, so it'd make for a better commit history.

ilya-biryukov commented on clang/lib/Basic/Warnings.cpp:
```diff
@@ -70,6 +169,14 @@ void clang::ProcessWarningOptions(DiagnosticsEngine &Diags,
 
+  if (!Opts.SuppressionMappingsFile.empty()) {
+    if (auto Buf = VFS.getBufferForFile(Opts.SuppressionMappingsFile)) {
+      parseSuppressionMappings(**Buf, Diags);
```
NIT: maybe make this return suppression mappings rather then setting them?
It would make the function interface more explicit (memory buffer as input, mapping as output) and would allow to move error handling to the client (no need to pass `Diags` as parameter).

NIT2: we could name the type for the lambda that we return (via typedef or through a real struct) to make these changes easier. I feel that right now the fact that the type is anonymous stops us from writing things like `Expected<ParsedMapping>`.

ilya-biryukov commented on clang/lib/Frontend/CompilerInstance.cpp:
```diff
@@ -332,19 +334,22 @@ static void SetupSerializedDiagnostics(DiagnosticOptions *DiagOpts,
   IntrusiveRefCntPtr<DiagnosticsEngine>
       Diags(new DiagnosticsEngine(DiagID, Opts));
 
+  if (!VFS)
```
Are we worried that we might forget to pass the VFS here and accidentally get `RealFS` when the clients intend to use their own?

These things have definitely happened in the past, and they are really expensive to debug.
I know it's tedious to change all the call sites, but I suggest considering to add this as a required parameter and move the choice of a VFS to call sites.
It's a trade-off that will pay back with the fact that we won't need to be chasing those bugs where the file wasn't found in the future.

kadircet replied:
yes, that's a real concern. I am planning to address them in a followup as mentioned in https://github.com/llvm/llvm-project/issues/112518. as this patch was getting big, and changes here shouldn't upset any clients until they start using suppression mappings.

ilya-biryukov commented on llvm/include/llvm/Support/SpecialCaseList.h:
```diff
@@ -122,7 +122,6 @@ class SpecialCaseList {
     unsigned match(StringRef Query) const;
 
-  private:
     StringMap<std::pair<GlobPattern, unsigned>> Globs;
```
Is this intended or can it be kept private?

kadircet replied:
this is intended, as we don't want to just check a match, but rather find the longest match. hence we implement our own matching logic based on the globs. I didn't want to change behavior for existing callers as this is likely more expensive than what they're doing today. in addition to that we also need to post-process sections based on their line numbers to map diagnostic groups to individual diagnostics.

putting all of this into SpecialCaseList felt like an overkill, and since this is already an internal class to SpecialCaseList and inheritors, I just made the data non-private instead. If there's interest I can lift them from WarningsSpecialCaseList implementation to the base instead.

AaronBallman replied:
This can be made protected at least though, right?

AaronBallman replied:
The only places I see it used are within `class WarningsSpecialCaseList : public llvm::SpecialCaseList` -- what am I missing?

AaronBallman replied:
Oh, I missed the inner class context, that makes sense, thanks! It unfortunate that we're losing encapsulation with the change, but I can hold my nose.

kadircet replied:
thanks, and I'd like to address these in a follow-up refactoring, sent out https://discourse.llvm.org/t/refactoring-llvm-specialcaselist/82931 to see what people think about this.

ilya-biryukov reviewed (commented):
I have only done a quick pass, but still wanted to leave a few NITs.

kadircet commented on clang/include/clang/Basic/Diagnostic.h:
```diff
@@ -946,6 +952,13 @@ class DiagnosticsEngine : public RefCountedBase<DiagnosticsEngine> {
+  /// the file they occur in.
+  /// These take presumed locations into account, and can still be overriden by
+  /// clang-diagnostics pragmas.
+  void setDiagSuppressionMapping(decltype(DiagSuppressionMapping) Mapping);
```
makes sense, restricting the API by taking in a memory-buffer and performing the parsing in DiagnosticsEngine instead.

kadircet committed:
Rename diag-option && move parsing to DiagnosticsEngine

kadircet commented:
gentle ping @AaronBallman, I believe you were supportive of this approach in final version of [RFC](https://discourse.llvm.org/t/rfc-add-support-for-controlling-diagnostics-severities-at-file-level-granularity-through-command-line/81292/12?u=kadircet) but I'd like to make sure this agreement was implemented in an accurate manner.

So it'd be great if you can check the overall interaction in https://github.com/llvm/llvm-project/blob/61b963131975ca3f6536b38427e25024f8759c57/clang/test/Misc/warning-suppression-mappings.cpp and https://github.com/llvm/llvm-project/blob/61b963131975ca3f6536b38427e25024f8759c57/clang/test/Misc/Inputs/suppression-mapping.txt. If those look good, I am happy to settle on the details of the implementation with other reviewers.

bricknerb commented:
Add that change to clang/docs/ReleaseNotes.rst?

bricknerb commented on clang/include/clang/Basic/Diagnostic.h:
```diff
@@ -946,6 +953,24 @@ class DiagnosticsEngine : public RefCountedBase<DiagnosticsEngine> {
   }
 
+  /// Diagnostic suppression mappings can be used to ignore diagnostics based on
+  /// the file they occur in. Mapping file is expected to be a special case list
```
How about, instead "to ignore diagnostics based on the file they occur in", say "to avoid generating specific diagnostics in specific files" or perhaps "to suppress" or something similar?
I'm not sure "ignore diagnostics" clarifies the behavior.

bricknerb commented on clang/include/clang/Basic/Diagnostic.h:
```diff
@@ -946,6 +953,24 @@ class DiagnosticsEngine : public RefCountedBase<DiagnosticsEngine> {
+  ///   src:*clang/foo/bar/*
+  ///
+  /// Such a mappings file suppress all diagnostics produced by -Wunused in all
+  /// sources under `clang/` directory apart from `clang/foo/`. Diagnostics
```
Why does the example start with * ?

kadircet replied:
because filepaths are relative to execution root of clang, and not all projects/build systems run clang in the source root. so we can have prefixes like `../../` or `build/genfiles/` .

happy to drop that if you think that looks confusing.

bricknerb replied:
Would "*clang" catch "xclang" or does it assume "clang" must be a full path part?
Also, it would catch all clang directories, and not necessarily "the" clang/ directory.
It might be better to make this more explicit and not try to write it in a way that tries to catch everything.

bricknerb commented on clang/include/clang/Basic/DiagnosticOptions.h:
```diff
@@ -108,6 +108,9 @@ class DiagnosticOptions : public RefCountedBase<DiagnosticOptions>{
   /// The file to serialize diagnostics to (non-appending).
   std::string DiagnosticSerializationFile;
 
+  /// File that defines diagnostic suppression mappings.
```
Clarify this is a filename or a full file path.

bricknerb commented on clang/lib/Basic/Diagnostic.cpp:
```diff
@@ -477,6 +485,100 @@ void DiagnosticsEngine::setSeverityForAll(diag::Flavor Flavor,
+  void processSections(DiagnosticsEngine &Diags) {
+    // Drop the default section introduced by special case list, we only support
+    // exact diagnostic group names.
+    Sections.erase("*");
```
Can we make the default section not introduced by making this a configurable behavior in SpecialCaseList instead of introducing it and then removing it?

bricknerb commented on clang/lib/Basic/Diagnostic.cpp:
```diff
@@ -477,6 +485,100 @@ void DiagnosticsEngine::setSeverityForAll(diag::Flavor Flavor,
+    }
+    llvm::sort(LineAndSection);
+    static constexpr auto kFlavor = clang::diag::Flavor::WarningOrError;
+    for (const auto &[_, Entry] : LineAndSection) {
```
Can we use std::views::values() ?

kadircet replied:
llvm is compiled with c++17, so unfortunately not :/

bricknerb replied:
That's too bad, consider adding a FIXME if you think it's a good idea.

bricknerb commented on clang/lib/Basic/Diagnostic.cpp:
```diff
@@ -477,6 +485,100 @@ void DiagnosticsEngine::setSeverityForAll(diag::Flavor Flavor,
+            << !Suggestion.empty() << Suggestion;
+        continue;
+      }
+      for (auto D : GroupDiags)
```
Replacing auto with diag::kind seems more readable to me.

bricknerb commented on clang/lib/Basic/Diagnostic.cpp:
```diff
@@ -477,6 +485,100 @@ void DiagnosticsEngine::setSeverityForAll(diag::Flavor Flavor,
+    }
+    llvm::sort(LineAndSection);
+    static constexpr auto kFlavor = clang::diag::Flavor::WarningOrError;
+    for (const auto &[_, Entry] : LineAndSection) {
```
If we're iterating over LineAndSection, shouldn't Entry be renamed to Section? Or SectionEntry?

bricknerb commented on clang/lib/Basic/Diagnostic.cpp:
```diff
@@ -477,6 +485,100 @@ void DiagnosticsEngine::setSeverityForAll(diag::Flavor Flavor,
+    for (const auto &[_, Entry] : LineAndSection) {
+      SmallVector<diag::kind, 256> GroupDiags;
+      if (Diags.getDiagnosticIDs()->getDiagnosticsInGroup(
+              kFlavor, Entry->first(), GroupDiags)) {
```
Replacing first() with getKey() seems more readable.

bricknerb commented on clang/lib/Basic/Diagnostic.cpp:
```diff
@@ -477,6 +485,100 @@ void DiagnosticsEngine::setSeverityForAll(diag::Flavor Flavor,
+        StringRef Suggestion =
+            DiagnosticIDs::getNearestOption(kFlavor, Entry->first());
+        Diags.Report(diag::warn_unknown_diag_option)
+            << static_cast<unsigned>(kFlavor) << Entry->first()
```
Put Entry->first() in a variable since you use it in multiple places.

bricknerb commented on clang/lib/Basic/Diagnostic.cpp:
```diff
@@ -477,6 +485,100 @@ void DiagnosticsEngine::setSeverityForAll(diag::Flavor Flavor,
+              kFlavor, Entry->first(), GroupDiags)) {
+        StringRef Suggestion =
+            DiagnosticIDs::getNearestOption(kFlavor, Entry->first());
+        Diags.Report(diag::warn_unknown_diag_option)
```
Do we have an integration test for this use case?

kadircet replied:
we've got a unittest in https://github.com/llvm/llvm-project/pull/112517/files#diff-a8d03c054381627f238bdb8f7914daa4abebdae705d313dda026ce36eca0fe2fR224-R233. why do you think an integration test is also necessary ?

bricknerb replied:
Assuming we hide some functions, this is definitely something that should be fixed.
Also, if the API of the new class, because it inherits from the base class, is wider than necessary, it's also a good reason to refactor.

bricknerb replied:
Seems useful to test end-to-end including flags, reading files, etc.
Hard for me to estimate the cost of generating or maintaining this.

kadircet replied:
> Assuming we hide some functions, this is definitely something that should be fixed.
> Also, if the API of the new class, because it inherits from the base class, is wider than necessary, it's also a good reason to refactor.
> We can do it in a followup.

I guess this comment is rather about the inheritance vs composition of the new `WarningSpecialCaseList`. All the newly exposed APIs are already tested through public interfaces in https://github.com/llvm/llvm-project/pull/112517/files#diff-a8d03c054381627f238bdb8f7914daa4abebdae705d313dda026ce36eca0fe2f. Because this is all the functionality being introduced as well, I don't see the point in testing those separately.

Still happy to do the refactoring for exposing specialcaselist just as a parser though.

> Seems useful to test end-to-end including flags, reading files, etc.
> Hard for me to estimate the cost of generating or maintaining this.

There's some end-to-end [integration test](https://github.com/llvm/llvm-project/pull/112517/files#diff-0f0bd95286925c4909908958c1104decb1068507c0ef516382a779ce463bb83d) that ensures the flag is propagated all the way from driver to diagsengine. I'd rather keep rest of the functionality in unittests as they're much easier to reason about when they go wrong, and resort to integration tests when there are bits of logic that can't really be tested through a unittests (which isn't the case here).

bricknerb commented on clang/lib/Basic/Diagnostic.cpp:
```diff
@@ -477,6 +485,100 @@ void DiagnosticsEngine::setSeverityForAll(diag::Flavor Flavor,
+    if (Section == DiagToSection.end())
+      return false;
+    auto SrcEntries = Section->second->Entries.find("src");
+    if (SrcEntries == Section->second->Entries.end())
```
Put Section->second->Entries in a variable to avoid duplication.

bricknerb commented on clang/lib/Basic/Diagnostic.cpp:
```diff
@@ -477,6 +485,100 @@ void DiagnosticsEngine::setSeverityForAll(diag::Flavor Flavor,
+    // Find the longest glob pattern that matches FilePath. A positive match
+    // implies D should be suppressed for FilePath.
+    llvm::StringRef LongestMatch;
+    bool LongestWasNegative;
```
Initialize.

bricknerb commented on clang/lib/Basic/Diagnostic.cpp:
```diff
@@ -477,6 +485,100 @@ void DiagnosticsEngine::setSeverityForAll(diag::Flavor Flavor,
+    auto SrcEntries = Section->second->Entries.find("src");
+    if (SrcEntries == Section->second->Entries.end())
+      return false;
+    // Find the longest glob pattern that matches FilePath. A positive match
```
Wrap this logic in a function.

bricknerb commented on clang/lib/Basic/Diagnostic.cpp:
```diff
@@ -477,6 +485,100 @@ void DiagnosticsEngine::setSeverityForAll(diag::Flavor Flavor,
 }
 
+namespace {
+class WarningsSpecialCaseList : public llvm::SpecialCaseList {
```
Consider moving this class to a separate file.

kadircet replied:
i'd rather keep it here as I actually want to make sure this stays as an implementation detail of diagnostics, and not used by any other parts of clang.

bricknerb replied:
Do we need to keep all implementation details in a single file?
Shouldn't we try to avoid growing huge files so they're easier to read, maintain etc.?
If it's useful somewhere else, someone would just move it and use it I guess.

kadircet commented on clang/include/clang/Driver/Options.td:
```diff
@@ -8979,3 +8979,8 @@ def wasm_opt : Flag<["--"], "wasm-opt">,
+
+def warning_suppression_mappings_EQ : Joined<["--"],
+  "warning-suppression-mappings=">, Group<Diag_Group>,
+  HelpText<"File containing list of sources to suppress diagnostics for. See XXX for format">,
```
oops, i was planning to refer to clang docs initially, but then noticed that none of the existing command line flags references clang webpage hence decided not to break that convention, but forgot to drop this one.

kadircet commented on clang/lib/Basic/Diagnostic.cpp:
```diff
@@ -477,6 +485,100 @@ void DiagnosticsEngine::setSeverityForAll(diag::Flavor Flavor,
 }
 
+namespace {
+class WarningsSpecialCaseList : public llvm::SpecialCaseList {
```
unfortunately we need to use inheritance, as we need access to internals for implemeting some custom "special-case-list" matching logic.

current users of `SpecialCaseList` also go with a similar approach for adding some custom support on top of the parser. I don't feel comfortable changing/exposing all these interfaces directly from `SpecialCaseList`. Especially because it has some backward-compatibility gurantees around regexes vs globs etc, and I don't want to carry that complexity into this new use case.

I think in an ideal world `SpecialCaseList` should just expose its parser, and we'd use it here. I can perform such a refactoring if you feel strong about this.

bricknerb replied:
Yes, I think the current state is pretty bad so if you're willing to refactor (in a separate PR, could also be a followup), this would be great. I think WarningsSpecialCaseList::createOrDie() would be surprising and I assume future changes could be harder to make and/or make the API much less clear.

bricknerb replied:
Add a FIXME/TODO ?

kadircet committed:
Add docs and address comments

kadircet commented:
> clang/docs/ReleaseNotes.rst?

I was planning to add docs in a follow up once the UX is more mature, but I guess it doesn't hurt to do that here as well.

kadircet reviewed (commented):
thanks for the review @bricknerb, this is ready for next round

bricknerb commented on clang/lib/Basic/Diagnostic.cpp:
```diff
@@ -477,6 +486,109 @@ void DiagnosticsEngine::setSeverityForAll(diag::Flavor Flavor,
+  }
+
+  bool isDiagSuppressed(diag::kind D, llvm::StringRef FilePath) const {
+    auto Section = DiagToSection.find(D);
```
This is actually an iterator to a pair where the second value is a pointer to Section, which, AFAIU, is declared in http://google3/third_party/llvm/llvm-project/clang/include/clang/Basic/Attr.td;l=2983;rcl=686934080.
If we could move some of the autos in this function to be named types, it might be easier to follow.

kadircet replied:
hmm not sure how you ended up with a reference to that. but no, this is the [Section](https://github.com/llvm/llvm-project/blob/4b44639a4320f980b3c9fa3b96e911e0741f179c/llvm/include/llvm/Support/SpecialCaseList.h#L132) in question.

underlying type is a `DenseMapIterator<diag::kind, llvm::SpecialCaseList::Section*>` modulo some const qualifiers here and there, probably. i haven't really seen any code that actually spells out iterator types, and style guide is also encouraging use of auto in such cases, https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable. so i'd rather keep it as-is

bricknerb replied:
Thanks!
I see the benefit of using auto, but in this method it's really hard to follow the types and they don't seem "obvious from the context".
Consider introducing more variables with proper names (like one for Section->second or SrcEntries->second) which would not be auto.
Also, perhaps don't use auto for SrcEntries?

Anyways, "Section" could be rename to something that represents the fact this points to (kind, section).

bricknerb commented on clang/lib/Basic/Diagnostic.cpp:
```diff
@@ -477,6 +486,109 @@ void DiagnosticsEngine::setSeverityForAll(diag::Flavor Flavor,
+    return;
+  }
+  SCL->processSections(*this);
+  DiagSuppressionMapping = [SCL(std::move(SCL))](diag::kind K,
```
I see diag::Kind is often called Diag, but sometimes also called D or DiagId.
Let's go with Diag or DiagId unless we really need to use a one letter param name?
Also below.

bricknerb commented on clang/lib/Basic/DiagnosticIDs.cpp:
```diff
@@ -575,25 +575,42 @@ DiagnosticIDs::getDiagnosticSeverity(unsigned DiagID, SourceLocation Loc,
+  if (!Diag.hasSourceManager())
+    return Result;
+
+  const auto &SM = Diag.getSourceManager();
```
This is typically called SrcMgr and typically we keep the full name and don't use auto.

kadircet replied:
hmm not sure where that's coming from, but in my experience it's actually much more common to name this as `SM` and use `auto`. so i'd rather keep it as-is (happy to provide some pointers offline if need be).

bricknerb replied:
"SM" is common.
"auto" not that common.
"SourceMgr" is also relatively common, much more readable and fits better to the style guide.

AaronBallman replied:
`SM` is "fine" (a more descriptive name is better of course), but generally we only use `auto` when the type is spelled out in the assignment. In this case, `getSourceManager()` is pretty close to spelling out the type name and so I'm not really opposed to `auto`.

bricknerb commented on clang/lib/Basic/DiagnosticIDs.cpp:
```diff
@@ -575,25 +575,42 @@ DiagnosticIDs::getDiagnosticSeverity(unsigned DiagID, SourceLocation Loc,
+  // We also use presumed locations here to improve reproducibility for
+  // preprocessed inputs.
+  if (!Mapping.isPragma()) {
+    if (auto PLoc = SM.getPresumedLoc(Loc);
```
Consider avoiding auto here to make this easier to read and more consistent with other Clang code.

kadircet replied:
i think it's pretty LLVM-stylistic to omit type when the initializer has enough information about the type (e.g. getXXX methods usually return something of type XXX), so I don't think it'll be more consistent with rest of clang code, but if you feel strongly about it happy to do that.

bricknerb replied:
Consistency wise: At least for getPresumedLoc(), I found a single case of using auto and 30+ cases of using the type explicitly.

AaronBallman replied:
Same comment here as above -- `getPresumedLoc()` is close enough that `auto` is okay, but so is spelling out the type fully.

bricknerb commented on clang/lib/Basic/Warnings.cpp:
```diff
@@ -41,8 +46,10 @@ static void EmitUnknownDiagWarning(DiagnosticsEngine &Diags,
       << (Prefix.str() += std::string(Suggestion));
 }
 
+
```
Remove extra blank line?

bricknerb commented on clang/lib/Basic/Diagnostic.cpp:
```diff
@@ -477,6 +486,109 @@ void DiagnosticsEngine::setSeverityForAll(diag::Flavor Flavor,
+
+void DiagnosticsEngine::setDiagSuppressionMapping(llvm::MemoryBuffer &MB) {
+  std::string Err;
+  auto SCL = WarningsSpecialCaseList::create(MB, Err);
```
Variable names should be nouns.
https://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly

bricknerb replied:
Please apply this widely if possible.

kadircet replied:
again, i do think despite the coding standards saying that, code in practice is much more flexible than that and I think it actually has its positives. developer doesn't need to think thoroughly about naming when it isn't so useful.  for example the name of the SpecialCaseList object (IMO) doesn't really matter here, it's an implementation detail that isn't visible to any reader outside of these 5 lines. whereas "important" objects that are keeping track of some invariants/state as logic does something is still quite important (e.g. logic that finds longest match), hence such objects have more descriptive names (and its worth putting the effort).

i think the same applies across other parts of the codebase and this file. so i'd rather keep them like this. I am happy to make adjustments if you feel like there are bits of the logic that can benefit from more descriptive names though.

kadircet commented on clang/include/clang/Driver/Options.td:
```diff
@@ -8998,3 +8998,8 @@ def wasm_opt : Flag<["--"], "wasm-opt">,
+
+def warning_suppression_mappings_EQ : Joined<["--"],
+  "warning-suppression-mappings=">, Group<Diag_Group>,
+  HelpText<"File containing list of sources to suppress diagnostics for.">,
```
updated text a little. as i was saying in the previous thread, i don't feel comfortable just pointing at the docs as it needs to be version aware and none of the other command line arguments are doing it today.

kadircet committed:
update usersmanual && some renaming

bricknerb commented on clang/lib/Basic/Diagnostic.cpp:
```diff
@@ -477,6 +486,109 @@ void DiagnosticsEngine::setSeverityForAll(diag::Flavor Flavor,
+  }
+  SCL->processSections(*this);
+  DiagSuppressionMapping = [SCL(std::move(SCL))](diag::kind DiagId,
+                                                  llvm::StringRef Path) {
```
Indentation is a bit surprising here.

bricknerb commented on clang/include/clang/Basic/Diagnostic.h:
```diff
@@ -946,6 +953,27 @@ class DiagnosticsEngine : public RefCountedBase<DiagnosticsEngine> {
+  /// matched against the globs as-is.
+  /// These take presumed locations into account, and can still be overriden by
+  /// clang-diagnostics pragmas.
+  void setDiagSuppressionMapping(llvm::MemoryBuffer &MB);
```
Replace MB with what this buffer should contain.
Other places use "File", "Input", "InputBuffer", "Buffer" etc, which are all more readable than the type acronym.

bricknerb commented on clang/lib/Basic/Diagnostic.cpp:
```diff
@@ -477,6 +486,109 @@ void DiagnosticsEngine::setSeverityForAll(diag::Flavor Flavor,
+    LineAndSectionEntry.reserve(Sections.size());
+    for (const auto &Entry : Sections) {
+      LineAndSectionEntry.emplace_back(
+          Entry.second.SectionMatcher->Globs.at(Entry.first()).second, &Entry);
```
There are 3 first and second references, consider putting some in variables to make it easier to follow what this actually does.

bricknerb replied:
Are we confident Globs.at() will always succeed?

bricknerb commented on clang/lib/Basic/Diagnostic.cpp:
```diff
@@ -477,6 +486,109 @@ void DiagnosticsEngine::setSeverityForAll(diag::Flavor Flavor,
+            << !Suggestion.empty() << Suggestion;
+        continue;
+      }
+      for (diag::kind D : GroupDiags)
```
Replace D with Diag to be more consistent with other Clang code.

bricknerb commented on clang/lib/Basic/Diagnostic.cpp:
```diff
@@ -477,6 +486,109 @@ void DiagnosticsEngine::setSeverityForAll(diag::Flavor Flavor,
+    bool LongestIsPositive = false;
+    for (const auto &[Category, Matcher] : CategoriesToMatchers) {
+      bool IsPositive = Category != "emit";
+      for (const auto &[Pattern, Glob] : Matcher.Globs) {
```
Assuming we could have a lot of patterns, should we sort them by length to make sure the first match is the longest so we can break the loop?

kadircet replied:
> RE: "rest of the speical-case-list actually match entries in line order", 

sorry I think I was confused while writing this one. I was trying to say, rest of the SpecialCaseList users can observe which line matched first. Hence performing a re-ordering of patterns would result in behavioral changes for these users (I can't say if it'll be a regression or not, as I don't fully understand their use cases).

> do you mean that the behavior here is significantly different. Would this be confusing?

not sure what you mean here exactly. overall we don't have a different matching behavior in a "user-visible" manner. implementation-wise it's slightly different than other sanitizer-list matching, as we keep searching for longest matching glob, rather than stopping at first match. it's documented as such and I think is intuitive for users (that the most specific entry takes precedence).

bricknerb replied:
Sure, I didn't mean to suggest we should do it for other SpecialCaseList users.
We can limit sorting to this use case.

It's not just implementation, it's also behavior.
Two similar mapping files in two different use cases would match different patterns.
I wonder if it is required to have a different matching behavior or we could match by order so that users would just put their patterns in the order they want them to be matched.
If in other use cases the matching is by order, and we could have a pattern wider than another (foo/* wider than foo/bar/*), why is this use case different?

kadircet replied:
> Sure, I didn't mean to suggest we should do it for other SpecialCaseList users.
> We can limit sorting to this use case.

I am skeptical of that being a net-positive even for warning-suppression-mappings. In the common path, compiler doesn't emit warnings. Hence most of the compilations will just end up sorting this list, and never query it afterwards.

> I wonder if it is required to have a different matching behavior or we could match by order so that users would just put their patterns in the order they want them to be matched.

I'd rather give users less ways to misconfigure this infrastructure. but by the nature of the current behavior, users can get the exact same end result by just ensuring a subdirectory is mentioned before its parents in the list, eg:
```
foo/bar/*=emit
foo/*
```

> If in other use cases the matching is by order, and we could have a pattern wider than another (foo/* wider than foo/bar/*), why is this use case different?

other use cases only allow inclusion, they don't allow exclusions.

bricknerb replied:
Performance isn't my main concern here, just API consistency.
Given that there are no exclusions in other use cases, I think no inconsistency can happen, so we should be good.

bricknerb commented on clang/lib/Basic/Warnings.cpp:
```diff
@@ -70,6 +76,16 @@ void clang::ProcessWarningOptions(DiagnosticsEngine &Diags,
     Diags.setExtensionHandlingBehavior(diag::Severity::Ignored);
 
+  if (!Opts.DiagnosticSuppressionMappingsFile.empty()) {
+    if (auto Buf =
```
This is typically called File and sometimes called FileOrError

kadircet replied:
sure, i think it's a little bit unfortunate to spend so much effort in a variable that's scoped to 2 lines.

bricknerb commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -4493,6 +4493,13 @@ static void RenderDiagnosticsOptions(const Driver &D, const ArgList &Args,
+          Args.getLastArg(options::OPT_warning_suppression_mappings_EQ)) {
+    if (!D.getVFS().exists(A->getValue()))
+      D.Diag(clang::diag::err_drv_no_such_file) << A->getValue();
+    CmdArgs.push_back(Args.MakeArgString(A->getSpelling() + A->getValue()));
```
Consider putting A->getValue() in a variable to avoid duplication.

AaronBallman commented on clang/lib/Basic/Diagnostic.cpp:
```diff
@@ -477,6 +486,109 @@ void DiagnosticsEngine::setSeverityForAll(diag::Flavor Flavor,
+          Entry.second.SectionMatcher->Globs.at(Entry.first()).second, &Entry);
+    }
+    llvm::sort(LineAndSectionEntry);
+    static constexpr auto kFlavor = clang::diag::Flavor::WarningOrError;
```
```suggestion
    static constexpr auto Flavor = clang::diag::Flavor::WarningOrError;
```

AaronBallman commented on clang/lib/Basic/Diagnostic.cpp:
```diff
@@ -477,6 +486,109 @@ void DiagnosticsEngine::setSeverityForAll(diag::Flavor Flavor,
+    llvm::sort(LineAndSectionEntry);
+    static constexpr auto kFlavor = clang::diag::Flavor::WarningOrError;
+    for (const auto &[_, SectionEntry] : LineAndSectionEntry) {
+      SmallVector<diag::kind, 256> GroupDiags;
```
256 doesn't seem particularly small -- was there a reason for this value?

AaronBallman commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -4493,6 +4493,13 @@ static void RenderDiagnosticsOptions(const Driver &D, const ArgList &Args,
+
+  if (const Arg *A =
+          Args.getLastArg(options::OPT_warning_suppression_mappings_EQ)) {
+    if (!D.getVFS().exists(A->getValue()))
+      D.Diag(clang::diag::err_drv_no_such_file) << A->getValue();
```
I'm not super keen on this because it's a TOCTOU issue. We should issue this diagnostic if we're not able to open the file at the time we need to open it. That also covers the case where the file exists but doesn't have read permissions.

kadircet committed:
Address comments

- Add more names to intermediate variables.
- Rewording in docs

AaronBallman commented on clang/docs/UsersManual.rst:
```diff
@@ -1315,6 +1319,29 @@ with its corresponding `Wno-` option.
+  [unused]
+  src:foo/*
+
+  $ clang --warning-suppression-mappings=mapping.txt -Wunused foo/bar.cc
```
Ah, okay, I think there's a small tweak to the example and I'll be happy. I'll suggest it below.

AaronBallman commented on clang/docs/UsersManual.rst:
```diff
@@ -1315,6 +1319,32 @@ with its corresponding `Wno-` option.
+  $ clang --warning-suppression-mappings=mapping.txt -Wunused foo/bar.cc
+  # This compilation won't emit any unused findings for sources under foo/
+  # directory. But it'll still complain for all the other sources, e.g:
+  $ cat foo/bar.cc
+  #include "dir/include.h" // clang flags unused declarations here.
+  #include "foo/include.h" // but unused warnings under this source is omitted.
+
+
```
```suggestion
  $ cat foo/bar.cc
  #include "dir/include.h" // Clang flags unused declarations here
  #include "foo/include.h" // but unused warnings from this header are omitted
  #include "next_to_bar_cc.h" // as are unused warnings from this header file.
  // Further, unused warnings in the remainder of bar.cc are also omitted.


```

bricknerb commented on clang/lib/Basic/Diagnostic.cpp:
```diff
@@ -477,6 +486,118 @@ void DiagnosticsEngine::setSeverityForAll(diag::Flavor Flavor,
+        continue;
+      }
+      for (diag::kind Diag : GroupDiags)
+        DiagToSection[Diag] = &SectionEntry->getValue();
```
Consider adding a comment clarifying that we intentionally override previous values so that the latest entry will take precedence.

kadircet committed:
Add fixme and update usersmanual

bricknerb commented on clang/lib/Basic/DiagnosticIDs.cpp:
```diff
@@ -599,6 +600,19 @@ DiagnosticIDs::getDiagnosticSeverity(unsigned DiagID, SourceLocation Loc,
     return diag::Severity::Ignored;
 
+  // Apply suppression mappings if severity wasn't explicitly mapped with a
+  // clang-diagnostics pragma to ensure pragmas always take precedence over
```
What happens when a pragma explicitly enables a warning?

kadircet replied:
pragma takes precedence and we don't suppress via the mappings. do you think the comment needs rewording?

bricknerb replied:
Carefully reading the comment, it's clear.
We could make it easier perhaps.

// Clang-diagnostics pragmas always take precedence over suppression mapping.

The rest of the information in the comment can be put inside the first if().

bricknerb commented on clang/test/Misc/Inputs/suppression-mapping.txt:
Consider also testing the same warning belonging to multiple groups.

bricknerb commented on clang/test/Misc/warning-suppression-mappings.cpp:
Consider also testing pragmas.

kadircet replied:
this is already tested in https://github.com/llvm/llvm-project/pull/112517/files#diff-a8d03c054381627f238bdb8f7914daa4abebdae705d313dda026ce36eca0fe2fR327-R332

bricknerb replied:
Yes, I meant as an integration test.

kadircet replied:
sorry i am struggling to see the value in an integration test while we can (and do) test it in a unittest. can you elaborate ?

bricknerb replied:
Unit tests cover the code and Clang integration tests are useful to test a feature end-to-end.
The compiler logic that handles the code until it reaches this point might break the assumptions we have in the specific code (especially if we think about future changes to the logic).

bricknerb commented on clang/unittests/Basic/DiagnosticTest.cpp:
```diff
@@ -167,4 +176,159 @@ TEST(DiagnosticTest, storedDiagEmptyWarning) {
   Diags.Report(CaptureConsumer.StoredDiags.front());
 }
+
+class SuppressionMappingTest : public testing::Test {
```
Do you also test pragmas here?

kadircet replied:
yes, see https://github.com/llvm/llvm-project/pull/112517/files#diff-a8d03c054381627f238bdb8f7914daa4abebdae705d313dda026ce36eca0fe2fR327-R332

kadircet committed:
Address comments

- Move function bodies out of the anon namespaces
- Rename variables
- Add lit tests for daigs belonging to multiple groups

bricknerb commented on clang/docs/WarningSuppressionMappings.rst:
```diff
@@ -0,0 +1,90 @@
+Clang allows diagnostics to be configured at a translation-unit granularity.
+If a ``foo.cpp`` is compiled with ``-Wfoo``, all transitively included headers
+also need to be clean. Hence turning on new warnings in large codebases can be
+difficult today. Since it requires cleaning up all the existing warnings,
```
RE "today": If this documentation should describe what the feature is and how to use, I wouldn't refer to "today" as if we don't have the feature.

Also, instead of "difficult" I would just continue to the next sentence to say it requires cleaning ...

bricknerb commented on clang/include/clang/Frontend/CompilerInstance.h:
```diff
@@ -701,11 +702,10 @@ class CompilerInstance : public ModuleLoader {
+  static IntrusiveRefCntPtr<DiagnosticsEngine> createDiagnostics(
+      DiagnosticOptions *Opts, DiagnosticConsumer *Client = nullptr,
+      bool ShouldOwnClient = true, const CodeGenOptions *CodeGenOpts = nullptr,
+      IntrusiveRefCntPtr<llvm::vfs::FileSystem> VFS = nullptr);
```
Optional: Given that this already has 5 parameters, consider wrapping the params in a struct to make the calls easier.

bricknerb commented on clang/lib/Basic/Diagnostic.cpp:
```diff
@@ -477,6 +486,136 @@ void DiagnosticsEngine::setSeverityForAll(diag::Flavor Flavor,
+}
+
+void DiagnosticsEngine::setDiagSuppressionMapping(llvm::MemoryBuffer &Input) {
+  std::string Err;
```
It seems more common to avoid the short name.
```suggestion
  std::string Error;
```

bricknerb commented on clang/lib/Basic/Diagnostic.cpp:
```diff
@@ -477,6 +486,136 @@ void DiagnosticsEngine::setSeverityForAll(diag::Flavor Flavor,
+  if (SrcEntriesIt == EntityTypeToCategories.end())
+    return false;
+  const llvm::StringMap<llvm::SpecialCaseList::Matcher> &CategoriesToMatchers =
+      SrcEntriesIt->second;
```
How about replacing lines 585-589 with:

const llvm::StringMap<llvm::SpecialCaseList::Matcher> &CategoriesToMatchers = EntityTypeToCategories.lookup("src");

I believe this will just work.

bricknerb replied:
Thanks, makes sense!
Now I feel like adding this functionality :).

bricknerb commented on clang/lib/Basic/Diagnostic.cpp:
```diff
@@ -477,6 +486,136 @@ void DiagnosticsEngine::setSeverityForAll(diag::Flavor Flavor,
+  bool LongestIsPositive = false;
+  for (const auto &Entry : CategoriesToMatchers) {
+    llvm::StringRef Category = Entry.first();
+    const auto &Matcher = Entry.second;
```
Consider avoiding auto here since the type isn't trivial.

bricknerb commented on clang/unittests/Basic/DiagnosticTest.cpp:
```diff
@@ -167,4 +176,159 @@ TEST(DiagnosticTest, storedDiagEmptyWarning) {
+      llvm::makeIntrusiveRefCnt<llvm::vfs::InMemoryFileSystem>();
+  DiagnosticsEngine Diags{new DiagnosticIDs(), new DiagnosticOptions};
+
+  std::vector<StoredDiagnostic> takeDiags() {
```
Why move instead of returning const& ?

bricknerb commented on clang/unittests/Basic/DiagnosticTest.cpp:
```diff
@@ -167,4 +176,159 @@ TEST(DiagnosticTest, storedDiagEmptyWarning) {
+  Diags.getDiagnosticOptions().DiagnosticSuppressionMappingsFile = "foo.txt";
+  clang::ProcessWarningOptions(Diags, Diags.getDiagnosticOptions(), *FS);
+  EXPECT_THAT(takeDiags(), testing::ElementsAre(WithMessage(
+                               "no such file or directory: 'foo.txt'")));
```
Why do we only test the message and not also level and possibly other fields?

bricknerb replied:
We might care whether we propagate this as an error, warning etc (I guess we could change that).
In another case we use `WithMessage()` below we test `rr_drv_malformed_warning_suppression_mapping`, which I guess we should care about more than just the text.
For consistency and making sure we have coverage, I suggest checking all the interesting fields, and not just the text.

bricknerb commented on clang/unittests/Basic/DiagnosticTest.cpp:
```diff
@@ -167,4 +176,159 @@ TEST(DiagnosticTest, storedDiagEmptyWarning) {
+}
+
+TEST_F(SuppressionMappingTest, SuppressesGroup) {
+  llvm::StringLiteral SuppressionMappingFile = R"txt(
```
Nit: Consider removing the "txt" as it has no effect here.
Seems like other use cases of R"()" in llvm don't set this.

bricknerb commented on clang/unittests/Basic/DiagnosticTest.cpp:
```diff
@@ -167,4 +176,159 @@ TEST(DiagnosticTest, storedDiagEmptyWarning) {
+  FS->addFile("foo.txt", /*ModificationTime=*/{},
+              llvm::MemoryBuffer::getMemBuffer("asdf", "foo.txt"));
+  clang::ProcessWarningOptions(Diags, Diags.getDiagnosticOptions(), *FS);
+  EXPECT_THAT(takeDiags(), testing::ElementsAre(WithMessage(
```
Consider adding using for these:
```
using testing::ElementsAre;
using testing::IsEmpty;
```

It seems like the common practice in Clang's unit tests.

bricknerb commented on clang/unittests/Basic/DiagnosticTest.cpp:
```diff
@@ -167,4 +176,159 @@ TEST(DiagnosticTest, storedDiagEmptyWarning) {
+  EXPECT_FALSE(Diags.isSuppressedViaMapping(diag::warn_deprecated, "foo.cpp"));
+}
+
+TEST_F(SuppressionMappingTest, ExclusionsWork) {
```
What do you mean by "Work"?
Would "Exclusion" be enough? Or "ExclusionSuccess"?

bricknerb commented on clang/unittests/Basic/DiagnosticTest.cpp:
```diff
@@ -167,4 +176,159 @@ TEST(DiagnosticTest, storedDiagEmptyWarning) {
+                                           "clang/lib/Sema/foo.h"));
+}
+
+TEST_F(SuppressionMappingTest, IsIgnored) {
```
Did we change the logic of isIgnored that we need to add specific tests for it here?

I think most of the tests should be for isSuppressedViaMapping() and we should have a few tests for ProcessWarningOptions() just to make the new specific logic there works (covering DiagnosticSuppressionMappingsFile empty, getBufferForFile() failing (with and without ReportDiags) and success.
I think it should also be clear from the test name which function (what API) is tested.

kadircet replied:
i modified `DiagnosticIDs::getDiagnosticSeverity` and it is an implementation detail of the public `isIgnored` API. hence yes, this patch changes the observable behavior of `isIgnored`.

> I think most of the tests should be for isSuppressedViaMapping() and we should have a few tests for ProcessWarningOptions() just to make the new specific logic there works (covering DiagnosticSuppressionMappingsFile empty, getBufferForFile() failing (with and without ReportDiags) and success.
> I think it should also be clear from the test name which function (what API) is tested.

I believe that's what all the previous tests in this file is doing already;

> we should have a few tests for ProcessWarningOptions() just to make the new specific logic there works (covering DiagnosticSuppressionMappingsFile empty, getBufferForFile() failing (with and without ReportDiags) and success.
- MissingMappingFile
- MalformedFile
- UnknownDiagNAme

> I think most of the tests should be for isSuppressedViaMapping()

Tests that make sure "behavior" of suppression mapping is WAI, with names explaining the behaviours:
- SuppressesGroup
- EmitCategoryIsExcluded
- LongestMatchWins
- IsIgnored

`IsIgnored` tests behavior in presence of presumedlocs and diagnostic-pragmas. hence it needs to test it through a source-location based API.

bricknerb commented on clang/unittests/Basic/DiagnosticTest.cpp:
```diff
@@ -167,4 +176,159 @@ TEST(DiagnosticTest, storedDiagEmptyWarning) {
+  Diags.getDiagnosticOptions().Warnings = {"unused"};
+  FS->addFile("foo.txt", /*ModificationTime=*/{},
+              llvm::MemoryBuffer::getMemBuffer(SuppressionMappingFile));
+  clang::ProcessWarningOptions(Diags, Diags.getDiagnosticOptions(), *FS);
```
Why do we test this in this test if we want to test isIgnored()?

kadircet replied:
it's not for testing the logic, but rather to make sure test-data is valid. turning it into an assertion instead. rest of this test assumes we successfully parsed the suppression mapping.

bricknerb replied:
Assert makes sense.

bricknerb reviewed (changes requested):
Note that there are a few open comments from previous iterations.

kadircet committed:
Add WarningSuppressionMappings to clang-root toctree

kadircet committed:
Reword comments/docs

kadircet reviewed (commented):
> Note that there are a few open comments from previous iterations.

I tried to go over them, but github UI was ... suboptimal. So PLMK if you got some that aren't addressed in this revision as well.

bricknerb commented on clang/unittests/Basic/DiagnosticTest.cpp:
```diff
@@ -167,4 +179,155 @@ TEST(DiagnosticTest, storedDiagEmptyWarning) {
+      llvm::makeIntrusiveRefCnt<llvm::vfs::InMemoryFileSystem>();
+  DiagnosticsEngine Diags{new DiagnosticIDs(), new DiagnosticOptions};
+
+  llvm::ArrayRef<StoredDiagnostic> takeDiags() {
```
Rename to getDiags()?
Take suggests they're no longer there.

bricknerb commented on clang/docs/WarningSuppressionMappings.rst:
```diff
@@ -0,0 +1,90 @@
+Users can also use the ``emit`` category to exclude a subdirectory from
+suppression.
+Source files are matched against these globs either as paths relative to th
+current working directory, or as absolute paths.
```
Careful reading it again this is ok.
If you want to emphasize it better, you can make it a list and/or emphasize the either.
... against these globs **either**:
* as ...
* as ...

bricknerb commented on clang/lib/Basic/Diagnostic.cpp:
```diff
@@ -477,6 +486,136 @@ void DiagnosticsEngine::setSeverityForAll(diag::Flavor Flavor,
+  if (SrcEntriesIt == EntityTypeToCategories.end())
+    return false;
+  const llvm::StringMap<llvm::SpecialCaseList::Matcher> &CategoriesToMatchers =
+      SrcEntriesIt->second;
```
I suggest using getValue() and getKey() instead of second and first throughout the code when possible, since it's more readable.

kadircet committed:
address comments

kadircet commented:
@AaronBallman merging this to main now. PLMK if you have any concerns, I am happy to address them post-submit (or revert if need be)

Michael137 commented:
FYI, looks like this is causing following LLDB builds to fail: https://green.lab.llvm.org/job/llvm.org/view/LLDB/job/as-lldb-cmake/15106/execution/node/97/log/
```
Unresolved Tests (17):
  lldb-api :: commands/expression/import-std-module/array/TestArrayFromStdModule.py
  lldb-api :: commands/expression/import-std-module/deque-basic/TestDequeFromStdModule.py
  lldb-api :: commands/expression/import-std-module/deque-dbg-info-content/TestDbgInfoContentDequeFromStdModule.py
  lldb-api :: commands/expression/import-std-module/forward_list-dbg-info-content/TestDbgInfoContentForwardListFromStdModule.py
  lldb-api :: commands/expression/import-std-module/forward_list/TestForwardListFromStdModule.py
  lldb-api :: commands/expression/import-std-module/iterator/TestIteratorFromStdModule.py
  lldb-api :: commands/expression/import-std-module/list-dbg-info-content/TestDbgInfoContentListFromStdModule.py
  lldb-api :: commands/expression/import-std-module/list/TestListFromStdModule.py
  lldb-api :: commands/expression/import-std-module/non-module-type-separation/TestNonModuleTypeSeparation.py
  lldb-api :: commands/expression/import-std-module/retry-with-std-module/TestRetryWithStdModule.py
  lldb-api :: commands/expression/import-std-module/shared_ptr-dbg-info-content/TestSharedPtrDbgInfoContentFromStdModule.py
  lldb-api :: commands/expression/import-std-module/shared_ptr/TestSharedPtrFromStdModule.py
  lldb-api :: commands/expression/import-std-module/unique_ptr-dbg-info-content/TestUniquePtrDbgInfoContent.py
  lldb-api :: commands/expression/import-std-module/unique_ptr/TestUniquePtrFromStdModule.py
  lldb-api :: commands/expression/import-std-module/vector-of-vectors/TestVectorOfVectorsFromStdModule.py
  lldb-api :: commands/expression/import-std-module/weak_ptr-dbg-info-content/TestDbgInfoContentWeakPtrFromStdModule.py
  lldb-api :: commands/expression/import-std-module/weak_ptr/TestWeakPtrFromStdModule.py
```

Example assertion:
```
Assertion failed: (0 && "Invalid SLocOffset or bad function choice"), function getFileIDLoaded, file SourceManager.cpp, line 867.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
``

AaronBallman commented on clang/docs/UsersManual.rst:
```diff
@@ -1315,6 +1319,34 @@ with its corresponding `Wno-` option.
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Warning suppression mappings enable users to suppress Clang's diagnostics in a
+per-file granular manner. Enabling enforcement of diagnostics in specific parts
```
Add comma: per-file, granular manner
Reword: Enabling enforcement of -> This allows enforcing

AaronBallman commented on clang/docs/UsersManual.rst:
```diff
@@ -1315,6 +1319,34 @@ with its corresponding `Wno-` option.
+
+Warning suppression mappings enable users to suppress Clang's diagnostics in a
+per-file granular manner. Enabling enforcement of diagnostics in specific parts
+of the project, even if there are violations in some headers.
```
Remove comma: project even if there
Reword: some headers -> some headers which cannot be modified.

kadircet replied:
> Reword: some headers -> some headers which cannot be modified.

But this covers headers even inside the project (i.e. the incremental cleanup of a new warning). So I feel like `which cannot be modified.` adds a non-existing limitation.

AaronBallman replied:
That's true, but I think we want to guide users towards fixing diagnostics rather than suppressing them. Users will use any suppression mechanisms we provide regardless of how we document it, so I don't think it's harmful to imply this is more about cases where you can't modify the code. But I also don't feel strongly. WDYT?

AaronBallman commented on clang/docs/WarningSuppressionMappings.rst:
```diff
@@ -0,0 +1,97 @@
+Introduction
+============
+
+Warning suppression mappings enable users to suppress Clang's diagnostics in a
+per-file granular manner. Enabling enforcement of diagnostics in specific parts
+of the project, even if there are violations in some headers.
```
Same suggestions here as above.

AaronBallman commented on clang/docs/WarningSuppressionMappings.rst:
```diff
@@ -0,0 +1,97 @@
+
+Clang allows diagnostics to be configured at a translation-unit granularity.
+If a ``foo.cpp`` is compiled with ``-Wfoo``, all transitively included headers
+also need to be clean. Hence turning on new warnings in large codebases requires
```
Add comma: Hence, turning on

AaronBallman commented on clang/lib/Basic/Diagnostic.cpp:
```diff
@@ -477,6 +486,136 @@ void DiagnosticsEngine::setSeverityForAll(diag::Flavor Flavor,
+  std::string Error;
+  auto WarningSuppressionList = WarningsSpecialCaseList::create(Input, Error);
+  if (!WarningSuppressionList) {
+    Report(diag::err_drv_malformed_warning_suppression_mapping)
```
Can you add a FIXME comment here: it would be better to not use `Error` here as a string but instead use `%select` in the diagnostic message; combining arbitrary strings makes it harder when folks are localizing diagnostics.

AaronBallman commented on clang/include/clang/Driver/Options.td:
```diff
@@ -9017,3 +9017,8 @@ def wasm_opt : Flag<["--"], "wasm-opt">,
   HelpText<"Enable the wasm-opt optimizer (default)">,
   MarshallingInfoNegativeFlag<LangOpts<"NoWasmOpt">>;
+
+def warning_suppression_mappings_EQ : Joined<["--"],
```
This should move elsewhere in Options.td, it's currently listed under `clang-dxc` options.

Also, should this be exposed in clang-cl as well? What about flang? CC @zmodem @rnk @MaskRay @jansvoboda11 for opinions

kadircet replied:
> This should move elsewhere in Options.td, it's currently listed under clang-dxc options.

Oh didn't notice we actually had "lexical" sections in the doc (AFAICT it doesn't have any semantic affect afterwards, flag is still only visible for cc1 and clang-driver).

> Also, should this be exposed in clang-cl as well? What about flang? CC @zmodem @rnk @MaskRay @jansvoboda11 for opinions

Going to carry this discussion into RFC to make sure it doesn't get lost. https://discourse.llvm.org/t/rfc-add-support-for-controlling-diagnostics-severities-at-file-level-granularity-through-command-line/81292/21?u=kadircet

AaronBallman reviewed (commented):
> @AaronBallman merging this to main now. PLMK if you have any concerns, I am happy to address them post-submit (or revert if need be)

Sorry, there was a US holiday on Monday and I was mostly out on Friday, so that's why there was a delay in review. In the future, I would recommend waiting before landing significant changes like these unless the review is accepted by a maintainer or someone known to be familiar with that area of the code. That reduces pressure on reviewers and allows folks to be sick/take vacations/etc without things landing out from under them. As it turns out, I can no longer leave suggested changes on the PR because it was already merged (thanks, GitHub....), so my review comments will be somewhat awkward.

kadircet commented:
> FYI, looks like this is causing following LLDB tests to fail: https://green.lab.llvm.org/job/llvm.org/view/LLDB/job/as-lldb-cmake/15106/execution/node/97/log/
> 
> ```
> Unresolved Tests (17):
>   lldb-api :: commands/expression/import-std-module/array/TestArrayFromStdModule.py
>   lldb-api :: commands/expression/import-std-module/deque-basic/TestDequeFromStdModule.py
>   lldb-api :: commands/expression/import-std-module/deque-dbg-info-content/TestDbgInfoContentDequeFromStdModule.py
>   lldb-api :: commands/expression/import-std-module/forward_list-dbg-info-content/TestDbgInfoContentForwardListFromStdModule.py
>   lldb-api :: commands/expression/import-std-module/forward_list/TestForwardListFromStdModule.py
>   lldb-api :: commands/expression/import-std-module/iterator/TestIteratorFromStdModule.py
>   lldb-api :: commands/expression/import-std-module/list-dbg-info-content/TestDbgInfoContentListFromStdModule.py
>   lldb-api :: commands/expression/import-std-module/list/TestListFromStdModule.py
>   lldb-api :: commands/expression/import-std-module/non-module-type-separation/TestNonModuleTypeSeparation.py
>   lldb-api :: commands/expression/import-std-module/retry-with-std-module/TestRetryWithStdModule.py
>   lldb-api :: commands/expression/import-std-module/shared_ptr-dbg-info-content/TestSharedPtrDbgInfoContentFromStdModule.py
>   lldb-api :: commands/expression/import-std-module/shared_ptr/TestSharedPtrFromStdModule.py
>   lldb-api :: commands/expression/import-std-module/unique_ptr-dbg-info-content/TestUniquePtrDbgInfoContent.py
>   lldb-api :: commands/expression/import-std-module/unique_ptr/TestUniquePtrFromStdModule.py
>   lldb-api :: commands/expression/import-std-module/vector-of-vectors/TestVectorOfVectorsFromStdModule.py
>   lldb-api :: commands/expression/import-std-module/weak_ptr-dbg-info-content/TestDbgInfoContentWeakPtrFromStdModule.py
>   lldb-api :: commands/expression/import-std-module/weak_ptr/TestWeakPtrFromStdModule.py
> ```
> 
> Example assertion:
> 
> ```
> Assertion failed: (0 && "Invalid SLocOffset or bad function choice"), function getFileIDLoaded, file SourceManager.cpp, line 867.
> PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
> ``
> ```

FWIW, I am still trying to reproduce these failures. It seems to be hard on a linux machine as tests are all disabled with a link to https://discourse.llvm.org/t/lldb-test-failures-on-linux/80095.

nikic commented:
It looks like this increases time to build clang by 0.5% (see affected files here: https://llvm-compile-time-tracker.com/compare_clang.php?from=e385e0d3e71e17da0b2023f480259c95923707bd&to=41e3919ded78d8870f7c95e9181c7f7e29aa3cc4&stat=instructions%3Au&sortBy=interestingness)

This usually indicates that you are including something expensive in a core header -- it would be good to check whether it can be avoided.

nikic commented:
Though looking at that file list in more details, this also affects files outside clang, so I think it might be adding some compile-time overhead that is only visible on specific workloads. E.g. `LLVMSupport.dir/UnicodeNameToCodepointGenerated.cpp` has a 4.5% increase.

Michael137 commented:
> > FYI, looks like this is causing following LLDB tests to fail: https://green.lab.llvm.org/job/llvm.org/view/LLDB/job/as-lldb-cmake/15106/execution/node/97/log/
> > ```
> > Unresolved Tests (17):
> >   lldb-api :: commands/expression/import-std-module/array/TestArrayFromStdModule.py
> >   lldb-api :: commands/expression/import-std-module/deque-basic/TestDequeFromStdModule.py
> >   lldb-api :: commands/expression/import-std-module/deque-dbg-info-content/TestDbgInfoContentDequeFromStdModule.py
> >   lldb-api :: commands/expression/import-std-module/forward_list-dbg-info-content/TestDbgInfoContentForwardListFromStdModule.py
> >   lldb-api :: commands/expression/import-std-module/forward_list/TestForwardListFromStdModule.py
> >   lldb-api :: commands/expression/import-std-module/iterator/TestIteratorFromStdModule.py
> >   lldb-api :: commands/expression/import-std-module/list-dbg-info-content/TestDbgInfoContentListFromStdModule.py
> >   lldb-api :: commands/expression/import-std-module/list/TestListFromStdModule.py
> >   lldb-api :: commands/expression/import-std-module/non-module-type-separation/TestNonModuleTypeSeparation.py
> >   lldb-api :: commands/expression/import-std-module/retry-with-std-module/TestRetryWithStdModule.py
> >   lldb-api :: commands/expression/import-std-module/shared_ptr-dbg-info-content/TestSharedPtrDbgInfoContentFromStdModule.py
> >   lldb-api :: commands/expression/import-std-module/shared_ptr/TestSharedPtrFromStdModule.py
> >   lldb-api :: commands/expression/import-std-module/unique_ptr-dbg-info-content/TestUniquePtrDbgInfoContent.py
> >   lldb-api :: commands/expression/import-std-module/unique_ptr/TestUniquePtrFromStdModule.py
> >   lldb-api :: commands/expression/import-std-module/vector-of-vectors/TestVectorOfVectorsFromStdModule.py
> >   lldb-api :: commands/expression/import-std-module/weak_ptr-dbg-info-content/TestDbgInfoContentWeakPtrFromStdModule.py
> >   lldb-api :: commands/expression/import-std-module/weak_ptr/TestWeakPtrFromStdModule.py
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Example assertion:
> > ```
> > Assertion failed: (0 && "Invalid SLocOffset or bad function choice"), function getFileIDLoaded, file SourceManager.cpp, line 867.
> > PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
> > ``
> > ```
> 
> FWIW, I am still trying to reproduce these failures. It seems to be hard on a linux machine as tests are all disabled with a link to https://discourse.llvm.org/t/lldb-test-failures-on-linux/80095.

AFAIK only a few of these are skipped on Linux. E.g., `TestIteratorFromStdModule.py` should still reproduce. Happy to help you reproduce/narrow this down

Michael137 commented:
Hmm this one's a bit strange. Looks like the issue is in:
```
// Clang-diagnostics pragmas always take precedence over suppression mapping.
if (!Mapping.isPragma()) {                                                   
  // We also use presumed locations here to improve reproducibility for      
  // preprocessed inputs.                                                    
  if (PresumedLoc PLoc = SM.getPresumedLoc(Loc);                             
      PLoc.isValid() && Diag.isSuppressedViaMapping(                         
                            DiagID, llvm::sys::path::remove_leading_dotslash(
                                        PLoc.getFilename())))                
    return diag::Severity::Ignored;                                          
}                                                                            
```
The assert triggers inside `getPresumedLoc`. So doesn't look like it's an inherent issue with the patch. But it seems to be breaking assumptions that LLDB is making about these source locations

kadircet commented:
so https://github.com/llvm/llvm-project/commit/12e3ed8de8c6063b15916b3faf67c8c9cd17df1f should resolve both lldb buildbot breakage and compile time regression in clang.

that being said, I don't think https://github.com/llvm/llvm-project/commit/12e3ed8de8c6063b15916b3faf67c8c9cd17df1f
is the right fix for lldb breakage. I think this change reveals an actual bug in lldb's handling of modules, but I can't repro locally. I hit a similar issue with clang-repl during development, you can find the relevant fix in https://github.com/llvm/llvm-project/pull/112517/files#diff-23c4b4ecf706e1f0b594e365095a343594e3978b47991718782e1f76064f241a.

Basically I think LLDB is doing something similar, it's probably loading/building modules with one SourceManager and then it's trying to consume them using another. @Michael137, If you can provide more detailed instructions about how to set up a similar testing environment in linux (I couldn't get the std-modules bit to work), or if you can provide a detailed stack trace from assertion failure, I can try to dig deeper.

Michael137 commented:
> so [12e3ed8](https://github.com/llvm/llvm-project/commit/12e3ed8de8c6063b15916b3faf67c8c9cd17df1f) should resolve both lldb buildbot breakage and compile time regression in clang.
> 
> that being said, I don't think [12e3ed8](https://github.com/llvm/llvm-project/commit/12e3ed8de8c6063b15916b3faf67c8c9cd17df1f) is the right fix for lldb breakage. I think this change reveals an actual bug in lldb's handling of modules, but I can't repro locally. I hit a similar issue with clang-repl during development, you can find the relevant fix in https://github.com/llvm/llvm-project/pull/112517/files#diff-23c4b4ecf706e1f0b594e365095a343594e3978b47991718782e1f76064f241a.
> 
> Basically I think LLDB is doing something similar, it's probably loading/building modules with one SourceManager and then it's trying to consume them using another. @Michael137, If you can provide more detailed instructions about how to set up a similar testing environment in linux (I couldn't get the std-modules bit to work), or if you can provide a detailed stack trace from assertion failure, I can try to dig deeper.

Awesome thanks for the quick workaround. I agree, LLDB is doing something funky here. Example backtrace for the assertion:
```
frame #19: 0x000000012919b5a4 liblldb.20.0.0git.dylib`IsTemplateDeclCommonStructurallyEquivalent(Ctx=0x000000016fdeef40, D1=0x0000000107a49230, D2=0x0000000107a48dc0) at ASTStructuralEquivalence.cpp:2073:10
frame #18: 0x000000012919b278 liblldb.20.0.0git.dylib`IsStructurallyEquivalent(Context=0x000000016fdeef40, Params1=0x0000000107a49d98, Params2=0x0000000107a48f10) at ASTStructuralEquivalence.cpp:1964:7
frame #17: 0x000000012381b288 liblldb.20.0.0git.dylib`clang::DiagnosticBuilder::~DiagnosticBuilder(this=0x000000016fdeec80) at Diagnostic.h:1313:24
frame #16: 0x0000000123822bfc liblldb.20.0.0git.dylib`clang::DiagnosticBuilder::~DiagnosticBuilder(this=0x000000016fdeec80) at Diagnostic.h:1313:26
frame #15: 0x0000000123822c74 liblldb.20.0.0git.dylib`clang::DiagnosticBuilder::Emit(this=0x000000016fdeec80) at Diagnostic.h:1278:28
frame #14: 0x0000000129b98b58 liblldb.20.0.0git.dylib`clang::DiagnosticsEngine::EmitDiagnostic(this=0x0000000113ec4400, DB=0x000000016fdeec80, Force=false) at Diagnostic.cpp:658:15
frame #13: 0x0000000129b98be4 liblldb.20.0.0git.dylib`clang::DiagnosticsEngine::ProcessDiag(this=0x0000000113ec4400, DiagBuilder=0x000000016fdeec80) at Diagnostic.h:1034:19
frame #12: 0x0000000129baefc0 liblldb.20.0.0git.dylib`clang::DiagnosticIDs::ProcessDiag(this=0x0000600000af0d60, Diag=0x0000000113ec4400, DiagBuilder=0x000000016fdeec80) const at DiagnosticIDs.cpp:777:7
frame #11: 0x0000000129badb98 liblldb.20.0.0git.dylib`clang::DiagnosticIDs::getDiagnosticLevel(this=0x0000600000af0d60, DiagID=2351, Loc=(ID = 2093809277), Diag=0x0000000113ec4400) const at DiagnosticIDs.cpp:501:18
frame #10: 0x0000000129bae270 liblldb.20.0.0git.dylib`clang::DiagnosticIDs::getDiagnosticSeverity(this=0x0000600000af0d60, DiagID=2351, Loc=(ID = 2093809277), Diag=0x0000000113ec4400) const at DiagnosticIDs.cpp:607:31
frame #9: 0x0000000129c0797c liblldb.20.0.0git.dylib`clang::SourceManager::getPresumedLoc(this=0x0000000112f3f730, Loc=(ID = 2093809277), UseLineDirectives=true) const at SourceManager.cpp:1463:41
frame #8: 0x000000012646726c liblldb.20.0.0git.dylib`clang::SourceManager::getDecomposedExpansionLoc(this=0x0000000112f3f730, Loc=(ID = 2093809277)) const at SourceManager.h:1286:18
frame #7: 0x0000000123fd5410 liblldb.20.0.0git.dylib`clang::SourceManager::getFileID(this=0x0000000112f3f730, SpellingLoc=(ID = 2093809277)) const at SourceManager.h:1146:12
frame #6: 0x0000000123ffaa20 liblldb.20.0.0git.dylib`clang::SourceManager::getFileID(this=0x0000000112f3f730, SLocOffset=2093809277) const at SourceManager.h:1901:12
frame #5: 0x0000000129c05b80 liblldb.20.0.0git.dylib`clang::SourceManager::getFileIDSlow(this=0x0000000112f3f730, SLocOffset=2093809277) const at SourceManager.cpp:778:10
frame #4: 0x0000000129c05ed4 liblldb.20.0.0git.dylib`clang::SourceManager::getFileIDLoaded(this=0x0000000112f3f730, SLocOffset=2093809277) const at SourceManager.cpp:867:5
frame #3: 0x0000000183ef8c1c libsystem_c.dylib`__assert_rtn + 284
frame #2: 0x0000000183ef9908 libsystem_c.dylib`abort + 128
frame #1: 0x0000000183fecf50 libsystem_pthread.dylib`pthread_kill + 288
frame #0: 0x0000000183fb4600 libsystem_kernel.dylib`__pthread_kill + 8
(lldb) f
frame #4: 0x0000000129c05ed4 liblldb.20.0.0git.dylib`clang::SourceManager::getFileIDLoaded(this=0x0000000112f3f730, SLocOffset=2093809277) const at SourceManager.cpp:867:5
   857      LessIndex = MiddleIndex;
   858    }
   859  }
   860 
   861  /// Return the FileID for a SourceLocation with a high offset.
   862  ///
   863  /// This function knows that the SourceLocation is in a loaded buffer, not a
   864  /// local one.
   865  FileID SourceManager::getFileIDLoaded(SourceLocation::UIntTy SLocOffset) const {
   866    if (SLocOffset < CurrentLoadedOffset) {
-> 867      assert(0 && "Invalid SLocOffset or bad function choice");
   868      return FileID();
   869    }
   870 
   871    return FileID::get(ExternalSLocEntries->getSLocEntryID(SLocOffset));
   872  }
   873 
   874  SourceLocation SourceManager::
   875  getExpansionLocSlowCase(SourceLocation Loc) const {
   876    do {
   877      // Note: If Loc indicates an offset into a token that came from a macro
(lldb) v SLocOffset
(clang::SourceLocation::UIntTy) SLocOffset = 2093809277

(lldb) v CurrentLoadedOffset
(clang::SourceLocation::UIntTy) CurrentLoadedOffset = 2147483648

(lldb) p SourceManager::MaxLoadedOffset
(const clang::SourceLocation::UIntTy) 2147483648
```

Not sure if that's enough to go on though. Let me know what else could be helpful here.

>  it's probably loading/building modules with one SourceManager and then it's trying to consume them using another
Yea that sounds entirely plausible.

kadircet commented:
> Not sure if that's enough to go on though. Let me know what else could be helpful here.

Unfortunately it isn't, can you include all the frames? at least some call that originates from LLDB ?

nikic commented:
> so [12e3ed8](https://github.com/llvm/llvm-project/commit/12e3ed8de8c6063b15916b3faf67c8c9cd17df1f) should resolve both lldb buildbot breakage and compile time regression in clang.

Thank you, I can confirm that this fixes the compile-time regression (or at least most of it).

kadircet commented:
alright NVM, apparently that breakage was due to https://github.com/llvm/llvm-project/pull/107849. I'll still reland this tomorrow after addressing @AaronBallman's latest comments.

kadircet commented on clang/lib/Basic/Diagnostic.cpp:
```diff
@@ -477,6 +486,136 @@ void DiagnosticsEngine::setSeverityForAll(diag::Flavor Flavor,
+namespace {
+// FIXME: We should isolate the parser from SpecialCaseList and just use it
+// here.
+class WarningsSpecialCaseList : public llvm::SpecialCaseList {
```
As you well know these discrepancies are coming from https://github.com/llvm/llvm-project/blob/main/clang/include/clang/Basic/LLVM.h. It's really hard to know/remember what's available there and what isn't. Moreover the header only forward-declares, so at usage side one actually needs to include the actual header from `llvm/...`.

Hence can't do this for `SpecialCaseList` but doing it for others that I can. But I think in the long run it'd be interesting to see if people still find this intermediate header useful or not.

kadircet commented on clang/lib/Basic/Warnings.cpp:
```diff
@@ -23,10 +23,15 @@
+#include "llvm/Support/VirtualFileSystem.h"
 #include <cstring>
-#include <utility>
+#include <memory>
+#include <vector>
```
oops, I had more implementation here initially.
